### ISSUE TITLE
New Rule Idea: Null forgiving operator and conditional access should not be gratuitous/detrimental

### ISSUE THREAD

**Issue by gregory-paidis-sonarsource** [SonarSource/sonar-dotnet#7320]
There are two mechanisms to either check for null or ignore a null check:

- [Conditional access via (?)](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/member-access-operators#null-conditional-operators--and-)
- [Null-forgiving operator (!)](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/null-forgiving)

There are some cases that one of them might be either redundant, or actually making the program crash every time:

```csharp
string neverNull = "not null";
_ = neverNull?.ToString(); // Noncompliant: conditional access is redundant, always valid
_ = neverNull!.ToString(); // Noncompliant: null-forgiving is redundant, no reason to check

string alwaysNull = null;
_ = alwaysNull?.ToString(); // always evaluates to null. Most likely we should not raise here, as it prevents a NRE 
_ = alwaysNull!.ToString(); // Noncompliant: always NRE, operator is detrimental
```

It might be two rules as well, one for null-forgiving `!` and one for conditional access `?`.

**Comment by pavel-mikula-sonarsource** [comment]
S2583 and S2589 might take care of the `?` part after MMF-2563

**Comment by luizfbicalho** [comment]
Same issue is thacked in dotnet Roslyn 
https://github.com/dotnet/roslyn/issues/68267

**Comment by luizfbicalho** [comment]
Any information on this rule?

**Comment by luizfbicalho** [comment]
I created a roslyn analyzer to implement this, but its not working as expected, but if anyone is interested in add some thoughts [https://github.com/luizfbicalho/NullableAnalyzer](https://github.com/luizfbicalho/NullableAnalyzer)

**Comment by mary-georgiou-sonarsource** [comment]
After considering this rule we decided to specify the following:
`Conditional access should not be gratuitous.`

```csharp
string s = Test(true);
_ = s?.ToString();  // Noncompliant, you should remove ? and if you want to suppress a warning use !

string Test(bool condition) => condition ? "Hello": null;

```

**Comment by mary-georgiou-sonarsource** [comment]
One more update: 
We realized that the following rule:

> After considering this rule we decided to specify the following: `Conditional access should not be gratuitous.`
> 
> ```cs
> string s = Test(true);
> _ = s?.ToString();  // Noncompliant, you should remove ? and if you want to suppress a warning use !
> 
> string Test(bool condition) => condition ? "Hello": null;
> ```
> 

I overlapping with S2583/S2589 so we decided to not proceed with its implementation.
